### PR TITLE
Render Aware Scheduler Interface

### PR DESCRIPTION
**Pull request by runspired**
[Rendered](https://github.com/runspired/rfcs/blob/modernized-scheduler/text/0957-modernized-scheduler.md)

This RFC Proposes replacing `@ember/runloop` ([Backburner.js](https://github.com/BackburnerJS/backburner.js)) with an [interface](https://www.typescriptlang.org/docs/handbook/interfaces.html) for common scheduling needs.

The interface describes *intent* for when work should be performed in relation to the native event queues and render cycle of the browser. The details of *how* that work is scheduled and flushed are up to the specific scheduler implementation, allowing for experiments in this space.

Additionally, this RFC proposes deprecations and alterations to associated async primitives such as RSVP to support this exploration.

### CODE REVIEW

**Comment by kategengler** [comment]
Moving to the Exploring stage requires the consensus of the relevant teams. See :https://github.com/emberjs/rfcs#exploring I am removing the label.

File: text/0957-modernized-scheduler.md [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+  let render = scheduler.render();
+  requestAnimationFrame(() => { doExpensiveWork('after', 7)() });
+  await render;
+  doExpensiveWork('render', 2)();
+  await Promise.resolve();
+  doExpensiveWork('promise', 3)();
+  await scheduler.render();
+  doExpensiveWork('renderAgain', 4)();
+  await scheduler.layout();
+  bdoExpensiveWork('layout', 5)();
```
Comment (NullVoxPopuli) [inline text/0957-modernized-scheduler.md:301]:
```suggestion
  doExpensiveWork('layout', 5)();
```

File: text/0957-modernized-scheduler.md [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+Using this strategy, lets observe what happens when we use normal `async/await`
+patterns to schedule some work.
+
+**Example 2**
+```ts
+
+// puts some blocking named work into the performance profiler
+// to make it more obvious where the work was done and what work it was
+function doExpensiveWork(name, step) {
+  const fn = new Function(`return function ${name}() {const start = performance.now(); while (performance.now() - start < 10) {} console.log('${step}. ${name}');};`);
```
Comment (NullVoxPopuli) [inline text/0957-modernized-scheduler.md:286]:
```suggestion
  const fn = new Function(`
    return function ${name}() {
      const start = performance.now(); 
      while (performance.now() - start < 10) {} 
      console.log('${step}. ${name}');
    };`
  );
```

suggestion to avoid horizontal scrolling

**Comment by NullVoxPopuli** [comment]
What are the SSR impacts? (if we were to re-invest SSR in the most ideal way)
- what is settled?
- ignore idle? do we ignore it?
  -> flush idle queue?
  -> ensure flushed?
- how to test with idle?

------------

if you have a lot of timings / queues, how do you prevent folks from just arbitrarily adding things to whatever the "next queue" is?

----------

 how do we suggest / teach when folks will want to use one queue over others?

**Comment by wagenet** [comment]
Given that @ef4 and @wycats are generally in favor of this direction we agreed at the RFC review meeting to move to Exploring.

**Comment by wycats** [comment]
Thumbs up. We got into the details in the spec meeting and I think this is absolutely ready for Exploring.

**Comment by ef4** [comment]
Status update here: this is looking good, @runspired is leading experimentation in the implementation side as the next action before putting this forward as formal public API.

**Comment by runspired** [comment]
As an update: I ran AuditBoard's test suite replacing the RSVP flush with a native microtask instead of flushing in the runloop, and EmberData 4.12.x (which drops all RSVP and almost all runloop utilization). A few observations:

EmberData 4.12.x upgrade itself had a few timing things to work out, but it didn't take us too long to get through them despite ~12k tests, and they all fell into a bucket of bugs that wouldn't exist with the proposed scheduler.

RSVP: changing to flush with a microtask failed only 3 tests, all for an identical cause:
In a route afterModel we had this:

```
scheduleOnce('afterRender', this, () => {
  this.send('onSetBreadcrumb');
});
```

With RSVP flushing as a native microtask, this now runs after instead of before the completion handling of many promises associated with the transition. Were we to adopt this proposal though, this would have been a non-issue. It was also 1 shared cause for 3 test failures out of 12k tests that proved easy to fix. 

I next tried removing our reliance on `@ember/jquery` to avoid its wrapping of xhr resolutions into the runloop. We rely on this because we set `useFetch: false` in all of our adapters. Notably: the default for EmberData has for all of `4.x` been `useFetch: true`.

This resulted in dozens of test failures, possibly as many as a hundred. In debugging, it appears the reason for these failures is that tests that triggered async actions and ember-concurrency tasks were relying on the runloop initiated by jquery to count towards settledness. ember-concurrency itself makes widespread use of the runloop for most task related work. Combined, this means that `await settled()` and `await render()` and `await click()` and similar in tests capture a large body of async that might otherwise escape.

However, fixing this is as simple as ember-concurrency adding its own test waiters. EmberData already uses one, but there are some scenarios it intentionally does not cover (background requests). In these tests the cause was the request was being kicked off several promises deep from the triggering action and thus escaped the bounds of settled due to the missing EC waiter. The reason the jquery runloop usage papered over this was that it joined the next EC tick flush to the promise resolution in the test. Even if EC did not add a waiter, this is a class of issues that would not exist if we were to adopt the proposed RFC because the associated EmberData waiter would have been properly captured despite being multiple promises deep.

Current conclusions:

- I see no significant blocker to continuing to explore this RFC 
- The biggest risk seems to be to tests not app code, and most prominently in locations where EC is in use
- We should shore up EC before shipping a new scheduler to avoid friction there
- apps using `useFetch: false` should really stop doing so ;)

**Comment by ef4** [comment]
Feedback from RFC review discussion:
 - let's add detail to "how we teach this" explaining exactly what is deprecated and how it should appear in the deprecation guide and what people should be told about replacements
 - what's the experience in terms of when people will see deprecations, when should they flip the optional feature flag.
 - what is the rollout timeline? Which deprecations until 6.0? 7.0? 
 -

File: text/0957-modernized-scheduler.md [hunk 3]
```diff
@@ -0,0 +1,646 @@
...
+
+### Phases
+
+With this understanding of the lifecycle of a Frame, roughly speaking we start to see a few
+patterns for how we might want to schedule work in a way that seamlessly coordinates with the
+framework and other parts of our application.
+
+We divide this work into 6 conceptual phases:
+
+- `tasks` - work that updates reactive state
```
Comment (NullVoxPopuli) [inline text/0957-modernized-scheduler.md:173]: we may need a sync phase between `tasks` and `render`

this phase would be before we try rendering.

e.g.:
- sync runs before render
  - and after component construction
  - not in the render tracking frame / does not invalidate the render tracking frame
- must not run when the parent context is torn down

our purpose for on.sync is to create a mechanism for constructing a stateful object with a lifetime ("resource") that can use reactive data as part of its constructor. when the stateful data changes, the resource is cleaned up and reconstructed. also, when the parent lifetime is destroyed, the resource is cleaned up. and then finally, the resource object itself has a stable identity regardless of whether it has the internal resource has been synchronized or reset.

tl;dr: stable object on the outside, unstable object on the inside (which gets reset based on reactivity)

**Comment by runspired** [comment]
@MelSumner I looked, afterRender there could be refactored away in a few different ways but even if you left it as-is it would have a clean migration path to any of a number of the queues.

**Comment by NullVoxPopuli** [comment]
_an_ implementation https://github.com/emberjs/ember.js/pull/21493

**Comment by NullVoxPopuli** [comment]
We should move this forward. It's vital for our perf

File: text/0957-modernized-scheduler.md [hunk 4]
```diff
@@ -533,10 +615,7 @@ await layout();
 
 This phase is for work that needs to read DOM but does not require adjusting reactive state.
 
-Writing DOM during this phase will error in development.
```
Comment (runspired) [inline text/0957-modernized-scheduler.md]: why'd this constraint drop?

File: text/0957-modernized-scheduler.md [hunk 5]
```diff
@@ -584,63 +661,188 @@ import { idle } from '@ember/scheduler';
...
-> This section could also include prior art, that is, how other frameworks in the same domain have solved this problem.
+React (`useLayoutEffect` / `useEffect`), Vue (`nextTick` and watcher
+`flush: 'pre' | 'post' | 'sync'`), Svelte (`tick()`), and Angular's move to
+zoneless change detection all give user code a way to schedule relative to
+framework-owned render timing. What is distinctive here is that the phases
+are awaitable promises and the strategy is swappable.
 
 ## Unresolved questions
 
-Is `@ember/scheduler` the right name? Or is there utility in a non-Ember associated name?
```
Comment (runspired) [inline text/0957-modernized-scheduler.md]: I partially think we do want an agnostic name here - but also we thought that when we did backburner and we never did get backburner to be used beyond ember.

File: text/0957-modernized-scheduler.md (hunk 4 above)
Comment (NullVoxPopuli) [reply text/0957-modernized-scheduler.md]: nay, it was moved to the table. i debated whether or not this should remain in, and probably for clarity it should remain

**Comment by NullVoxPopuli** [comment]
Implementation: https://github.com/emberjs/ember.js/pull/21552


All additive, nothing to worry about with existing code

File: text/0957-modernized-scheduler.md [hunk 6]
```diff
@@ -0,0 +1,854 @@
...
+
+```ts
+import { idle } from '@ember/scheduler';
+
+// ...
+
+await idle();
+```
+
+### Migration Roadmap
```
Comment (NullVoxPopuli) [inline text/0957-modernized-scheduler.md:670]: TODO: 
- explicitly write scheduling RFC comes first, then deprecate RSVP/runloop (people need a thing to migrate to)

**Comment by NullVoxPopuli** [comment]
From RFC meeting:

migration plan needs more thoroughness:
- find a way to keep old addons from updating (provide a way to shim rsvp? and shim backburner and the runloop (using the new scheduler))
  - runloop re-implemented (via the optional-feature flag) to use the new scheduler
  - RSVP not used internally
    - same optional feature flag - Ember no longer uses rsvp
  - backburner
    - same optional feature flag - Ember no longer uses backburner


I think this enables the _fast_ implementation of rendering (via the optional feature flag) to land a lot sooner, ahead of deprecating the things

File: text/0957-modernized-scheduler.md [hunk 7]
```diff
@@ -0,0 +1,854 @@
...
+and the development-mode assertions cover misuse at the point it happens:
+scheduling with no registered strategy, and the phase-rule violations in
+the table above.
+
+### Existing users
+
+Existing users know this problem space through `@ember/runloop`; the
+deprecation guides for RFCs #1219 and #1220 teach by translation:
+
+| Today | After this RFC's roadmap |
```
Comment (NullVoxPopuli) [inline text/0957-modernized-scheduler.md:761]: TODO: spike a PR that makes the runloop use the scheduler (when the optional feature flag is on)

Comment (NullVoxPopuli) [reply text/0957-modernized-scheduler.md:761]: TODO:
- mention https://github.com/emberjs/rfcs/pull/1221

Also, add a column for "behavior when feature flag is enabled"

**Comment by alexraputa** [comment]
Thanks for working on this. I read through the current [#21552](https://github.com/emberjs/ember.js/pull/21552) implementation and the latest [#21520](https://github.com/emberjs/ember.js/pull/21520) spike, and had a few questions about the public scheduler semantics.

---

### 1. Is `Strategy` intended to describe phase timing rather than scheduling the work itself?

The RFC already says:

> the strategy has no knowledge of the work to be done

and the split in [#21552](https://github.com/emberjs/ember.js/pull/21552) seems to reinforce that - phase one exposes the interface without connecting it to Ember's renderer.

Would it be useful to make that distinction a little more explicit?

Something like:

> A `Strategy` controls when scheduler phase barriers become observable to callers, while the mechanism that determines what renderer work needs to happen is separate.

I think of the separation roughly like this:

```mermaid
flowchart LR
    State["Application / reactive state"]
    Renderer["Ember renderer"]
    DOM["DOM update"]

    Render["render()"]
    Layout["layout()"]
    Composite["composite()"]

    Strategy["Strategy"]

    State --> Renderer
    Renderer --> DOM
    DOM --> Render
    Render --> Layout
    Layout --> Composite

    Strategy -. "phase timing" .-> Render
    Strategy -. "phase timing" .-> Layout
    Strategy -. "phase timing" .-> Composite
```

Would that be a useful boundary to establish here, while leaving invalidation and renderer scheduling to the follow-up RFC?

---

### 2. What should `render()` mean when there is no pending render work?

Once Ember rendering is integrated with the scheduler, the renderer-related case seems fairly clear:

```ts
state.value = 123;

await render();

// DOM reflecting state.value should be observable here.
```

I'm less sure about the no-work case:

```ts
await render();
```

The current phase-one `FrameStrategy` in [#21552](https://github.com/emberjs/ember.js/pull/21552) isn't connected to the renderer yet, but requesting a phase when no frame exists creates or joins an upcoming frame:

[`FrameStrategy._phase()` / `_ensureFrame()` / `_scheduleFrame()`](https://github.com/emberjs/ember.js/blob/ff2ff075e8363864f3ccf9999e2c1a302f554bdb/packages/@ember/scheduler/strategy.ts#L119-L180)

The [#21520](https://github.com/emberjs/ember.js/pull/21520) spike explores the renderer-integrated version of this case. Its strategy can request a tick even when there is no reactive dirt, and it has a fallback when no renderer can provide that tick:

[`RenderClockStrategy` tick request / renderer tick handling](https://github.com/emberjs/ember.js/blob/c09efdc87933cc9cbd6fa2437239824b3eb4c878/packages/@ember/scheduler/strategy.ts#L76-L140)

Should calling `render()` itself establish an upcoming render-phase opportunity, even when there is currently no pending renderer work?

And is the intended semantic distinction roughly:

```text
render()
  = a render-phase barrier

render()
  != "render until the application is settled"
```

In other words, would a clean `await render()` still be expected to resolve at a render-phase opportunity rather than waiting for some unrelated future invalidation?

---

### 3. When renderer integration lands, should the scheduler phases and Ember rendering describe one logical render cycle?

This was the main question I had after comparing [#21552](https://github.com/emberjs/ember.js/pull/21552) with the current spike.

In [#21552](https://github.com/emberjs/ember.js/pull/21552), the default strategy owns its frame independently, which seems natural while the interface is intentionally not connected to Ember rendering:

[`FrameStrategy`](https://github.com/emberjs/ember.js/blob/ff2ff075e8363864f3ccf9999e2c1a302f554bdb/packages/@ember/scheduler/strategy.ts#L119-L180)

The [#21520](https://github.com/emberjs/ember.js/pull/21520) spike explores one possible end state where the renderer and scheduler share a clock.

The strategy accepts a tick requester and exposes `_onRendererTick()`:

[`RenderClockStrategy`](https://github.com/emberjs/ember.js/blob/c09efdc87933cc9cbd6fa2437239824b3eb4c878/packages/@ember/scheduler/strategy.ts#L76-L140)

The renderer connects itself as the tick source here:

[`schedulerStrategy._setTickRequester(...)`](https://github.com/emberjs/ember.js/blob/c09efdc87933cc9cbd6fa2437239824b3eb4c878/packages/@ember/-internals/glimmer/lib/base-renderer.ts#L211-L219)

and, after revalidation and the destroy drain, drives the phase windows here:

[`revalidate()` → `_onRendererTick()`](https://github.com/emberjs/ember.js/blob/c09efdc87933cc9cbd6fa2437239824b3eb4c878/packages/@ember/-internals/glimmer/lib/base-renderer.ts#L505-L517)

That experiment roughly gives us:

```mermaid
flowchart LR
    State["state mutation"]
    Work["Ember render work"]
    DOM["DOM updates + modifiers"]
    Render["render() phase"]
    Layout["layout() phase"]
    Composite["composite() phase"]

    State --> Work
    Work --> DOM
    DOM --> Render
    Render --> Layout
    Layout --> Composite
```

I don't think RFC necessarily needs to decide whether the renderer owns the clock, the strategy owns the clock, or how the underlying tick is scheduled.

But once the renderer is integrated, should these phases be understood as describing the **same logical render cycle** that produced the DOM state observed by `render()`?

Something like:

```text
Ember rendering timeline
        +
scheduler phase timeline
        =
one logical render cycle
```

Would that be a useful public invariant to establish here, while leaving the mechanism used to achieve it to the later async-rendering work?

---

### 4. How is `registerStrategy()` intended to interact with the future renderer integration?

I had two related questions here.

#### Is the strategy runtime-global or application-scoped?

The current [#21552](https://github.com/emberjs/ember.js/pull/21552) implementation keeps the registered strategy in module-level state:

[`registerStrategy()` / `getStrategy()`](https://github.com/emberjs/ember.js/blob/ff2ff075e8363864f3ccf9999e2c1a302f554bdb/packages/@ember/scheduler/index.ts#L113-L132)

while the API is described as registering it while defining the `Application`.

Is the intended ownership effectively one strategy per loaded Ember runtime, shared by applications/renderers in that runtime?

If so, would it be useful to state that explicitly?

#### Does a custom strategy also become the renderer's phase policy?

The public scheduler functions resolve their strategy through:

[`registeredStrategy ?? defaultStrategy`](https://github.com/emberjs/ember.js/blob/c09efdc87933cc9cbd6fa2437239824b3eb4c878/packages/@ember/scheduler/index.ts#L108-L125)

while the experimental renderer integration currently connects directly to the imported default `schedulerStrategy`:

[`schedulerStrategy._setTickRequester(...)`](https://github.com/emberjs/ember.js/blob/c09efdc87933cc9cbd6fa2437239824b3eb4c878/packages/@ember/-internals/glimmer/lib/base-renderer.ts#L211-L219)

That seems completely reasonable for a spike, and I'm not assuming that this split represents intended API semantics.

If an application eventually does this:

```ts
registerStrategy(customStrategy);
```

is that intended to replace the phase policy for both:

```text
public @ember/scheduler phase functions
```

and:

```text
future Ember renderer integration
```

or is `registerStrategy()` only intended to affect calls made through the public scheduler API?

Would this relationship be something useful to clarify in RFC, while leaving the actual renderer-to-strategy connection to the follow-up RFC?

---

### 5. Are the development-mode phase restrictions intended to be fully enforceable?

The phase table is useful:

| Phase | Write reactive state | Read DOM | Write DOM |
| --- | --- | --- | --- |
| `tasks` | ✅ | ✅ | ✅ |
| `render` | ✅ | ✅ | ❌ |
| `layout` | ❌ | ✅ | ❌ |
| `composite` | ❌ | ⚠️ discouraged | ✅ |
| `next` | ✅ | ✅ | ✅ |
| `idle` | ✅ | ✅ | ✅ |

The RFC says violations of the ❌ rules error in development mode.

For reactive state changes, I can see how Ember could observe that.

I'm less sure about arbitrary native DOM operations:

```ts
await layout();

element.style.width = '10px';
```

or:

```ts
await composite();

element.getBoundingClientRect();
```

Is the intention that these are semantic rules for the phases, with development assertions where Ember is actually able to observe a violation?

If so, would that distinction be useful to mention so the RFC doesn't imply that Ember necessarily intercepts arbitrary DOM reads and writes?

---

### Small question: should `MessageChannel` be categorized as a task rather than a microtask?

The Frames section currently groups `MessageChannel` with `queueMicrotask`, Promise reactions, and `MutationObserver`.

My understanding is that `MessagePort` messages are dispatched as tasks rather than microtasks:

https://html.spec.whatwg.org/multipage/web-messaging.html#dom-messageport-postmessage

This distinction also seems relevant to the later scheduler experiments, where `MessageChannel` is useful specifically as a task boundary.

Would it make sense to adjust that part of the description?

File: text/0957-modernized-scheduler.md (hunk 7 above)
Comment (NullVoxPopuli) [reply text/0957-modernized-scheduler.md:761]: Result: not possible because current behavior has synchronous flushing

Comment (NullVoxPopuli) [reply text/0957-modernized-scheduler.md:761]: Still seeing how much we can keep tho
